### DISCUSSION TITLE
Compiling multiple builds fails

### DISCUSSION THREAD

**Discussion by thermseekr** [arendst/Tasmota#22620] [Support request] [answered]
Hi there

I'm trying to create two different environments for compiling Tasmota following directions from [here](https://tasmota.github.io/docs/Compile-your-build/#defining-multiple-custom-firmwares). What I did so far:

In platformio_tasmota_cenv.ini:
```
[env:tasmota32-esp24v16ch]
extends                     = env:tasmota32_base
build_flags                 = ${env:tasmota32_base.build_flags}
                              -DFIRMWARE_ESP24V16CH

[env:tasmota32-crazy]
extends                     = env:tasmota32_base
build_flags                 = ${env:tasmota32_base.build_flags}
                              -DFIRMWARE_CRAZY

```

In user_config_override.h:
```
#ifdef FIRMWARE_ESP24V16CH

  #ifndef USE_MCP230xx_DRV
  #define MCP230xx_DRV
  #endif

  #define USER_TEMPLATE   "{\"NAME\":\"ESP-24V-16CH-V2\",\"GPIO\":[0,3200,0,3232,0,0,0,0,5536,10240,10240,0,0,0,5600,0,0,0,0,5568,0,0,0,0,0,0,0,0,608,640,0,0,0,0,0,0],\"FLAG\":0,\"BASE\":1}"

#endif

#ifdef FIRMWARE_CRAZY

  #ifdef USE_MCP230xx_DRV
  #undef MCP230xx_DRV
  #endif

  #define USER_TEMPLATE   "{\"NAME\":\"ESP-CRAZY\",\"GPIO\":[0,3200,0,3232,0,0,0,0,5536,0,0,0,0,0,5600,0,0,0,0,5568,0,0,0,0,0,0,0,0,608,640,0,0,0,0,0,0],\"FLAG\":0,\"BASE\":1}"

#endif

#ifdef MODULE
#undef MODULE
#endif
#define MODULE USER_MODULE

```

Upon compile I get the following:
```
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino: In function 'void ShutterUpdatePosition()':
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino:1305:13: error: 'Energy' was not declared in this scope; did you mean 'EnergyFmt'?
 1305 |         if (Energy->active_power[0] + Energy->active_power[1] < 1.0 && Shutter[i].time > 100){
      |             ^~~~~~
      |             EnergyFmt
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino: In function 'void CmndShutterSetup()':
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino:2097:19: error: 'ENERGY_MAX_PHASES' was not declared in this scope
 2097 |   float daily_kWh[ENERGY_MAX_PHASES];           // 123.123 kWh
      |                   ^~~~~~~~~~~~~~~~~
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino:2107:7: error: 'daily_kWh' was not declared in this scope
 2107 |       daily_kWh[i] = Energy->daily_kWh[i];
      |       ^~~~~~~~~
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino:2107:22: error: 'Energy' was not declared in this scope; did you mean 'EnergyFmt'?
 2107 |       daily_kWh[i] = Energy->daily_kWh[i];
      |                      ^~~~~~
      |                      EnergyFmt
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino:2110:9: error: 'Energy' was not declared in this scope; did you mean 'EnergyFmt'?
 2110 |     if (Energy->phase_count > 1) {
      |         ^~~~~~
      |         EnergyFmt
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino:2122:11: error: 'Energy' was not declared in this scope; did you mean 'EnergyFmt'?
 2122 |       if (Energy->phase_count > 1) {
      |           ^~~~~~
      |           EnergyFmt
C:/Users/tales.maschio/Documents/PlatformIO/Projects/Tasmota-14.3.0/tasmota/tasmota_xdrv_driver/xdrv_27_esp32_shutter.ino:2142:8: error: 'Energy' was not declared in this scope; did you mean 'EnergyFmt'?
 2142 |        Energy->daily_kWh[i] = daily_kWh[i];
      |        ^~~~~~
      |        EnergyFmt

```

I'm apparently missing some energy library. Can someone help me here?

EDIT: just double checked, I disabled (#undef) all references to energy metering via user_config_override.

Thanks
Tales

**Comment by thermseekr** [comment]
Good day everyone

I'm resuming trying to set up my environment for multiple builds. As I posted above, if I follow the instructions on the documentation the compilation fails, it looks like some energy related variables are not defined.

What I just found out, is that if I try to compile `env: tasmota32_base`, the compilation also fails. So, it makes sense that my builds will fail if they are based on `tasmota32_base`.

If I compile my builds from tasmota32 instead, all goes fine.

Does `tasmota32_base` depend on other definitions to be able to be compiled? I'd like to understand this better as I don't know if it's ok to compile custom builds from `tasmota32`.

Thanks
Tales

**Reply by sfromis** [reply, marked as answer]
Using `tasmota32_base` is only when you want to **fully** control what you want to enable, without helpful defaults. As a starting point, it is even suggested to base custom builds on `tasmota32`, unless you want to manage all details, like with "inspiration" from what you get with `tasmota32`. If you "just" want to enable some optional feature, the easy way is indeed `tasmota32` for a custom build.

**Reply by thermseekr** [reply]
Thank you @sfromis 
Compiling from `tasmota32` is indeed enough for my curent purposes.
Anyway, for the sake of knowledge, is there any documentation on how to use `tasmota32_base`?

**Reply by sfromis** [reply]
Not phrased like that. More generally, using build environments has a bit of examples/intro here:
https://tasmota.github.io/docs/Compile-your-build/#defining-multiple-custom-firmwares
Not "this is how to do it", more like to illustrate how the standard build process does it.
The point being that this can be a "starting point" to dive into the "real docs", code .....
